api/src/main/java/marquez/service/SearchService.java

```diff
@@ -0,0 +1,237 @@
+import org.opensearch.client.opensearch.core.IndexRequest;
+import org.opensearch.client.opensearch.core.SearchResponse;
+import org.opensearch.client.opensearch.core.search.BuiltinHighlighterType;
+import org.opensearch.client.opensearch.core.search.HighlighterType;
+import org.opensearch.client.transport.OpenSearchTransport;
+import org.opensearch.client.transport.endpoints.BooleanResponse;
+import org.opensearch.client.transport.rest_client.RestClientTransport;
+
+@Slf4j
+public class SearchService {
+
+  private final OpenSearchClient openSearchClient;
+
+  public SearchService(SearchConfig searchConfig) {
+    final HttpHost host =
+        new HttpHost(searchConfig.getHost(), searchConfig.getPort(), searchConfig.getScheme());
+    final BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+    credentialsProvider.setCredentials(
+        new AuthScope(host),
+        new UsernamePasswordCredentials(searchConfig.getUsername(), searchConfig.getPassword()));
+    final RestClient restClient =
+        RestClient.builder(host)
+            .setHttpClientConfigCallback(
+                httpClientBuilder ->
+                    httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider))
+            .build();
+
+    JacksonJsonpMapper jsonpMapper = new JacksonJsonpMapper();
+    // register JavaTimeModule to handle ZonedDateTime
+    jsonpMapper.objectMapper().registerModule(new JavaTimeModule());
+    final OpenSearchTransport transport = new RestClientTransport(restClient, jsonpMapper);
+    this.openSearchClient = new OpenSearchClient(transport);
+    BooleanResponse booleanResponse;
+    try {
+      booleanResponse = openSearchClient.ping();
+      log.info("OpenSearch Active: {}", booleanResponse.value());
+    } catch (IOException e) {
+      log.warn("Search not configured");
+    }
+  }
+
+  public OpenSearchClient getClient() {
+    return this.openSearchClient;
+  }
+
+  public SearchResponse<ObjectNode> searchDatasets(String query) throws IOException {
+    String[] fields = {
+      "run_id",
+      "name",
+      "namespace",
+      "facets.schema.fields.name",
+      "facets.schema.fields.type",
+      "facets.columnLineage.fields.*.inputFields.name",
+      "facets.columnLineage.fields.*.inputFields.namespace",
+      "facets.columnLineage.fields.*.inputFields.field",
+      "facets.columnLineage.fields.*.transformationDescription",
+      "facets.columnLineage.fields.*.transformationType"
+    };
+    return this.openSearchClient.search(
+        s ->
+            s.index("datasets")
+                .query(
+                    q ->
+                        q.multiMatch(
+                            m ->
+                                m.query(query)
+                                    .type(TextQueryType.PhrasePrefix)
+                                    .fields(Arrays.stream(fields).toList())
+                                    .operator(Operator.Or)))
```

Comment:
Minor: We'll want to support operators as part of the search query as mentioned [here](https://github.com/MarquezProject/marquez/pull/2834/files#r1704682610). But, not a blocker on the first pass.